hibernate-core/src/main/java/org/hibernate/hql/internal/ast/QueryTranslatorImpl.java

```diff
@@ -595,37 +595,47 @@ public void validateScrollability() throws HibernateException {
 	}
 
 	private StatementExecutor buildAppropriateStatementExecutor(HqlSqlWalker walker) {
-		final Statement statement = (Statement) walker.getAST();
 		if ( walker.getStatementType() == HqlSqlTokenTypes.DELETE ) {
 			final FromElement fromElement = walker.getFinalFromClause().getFromElement();
 			final Queryable persister = fromElement.getQueryable();
 			if ( persister.isMultiTable() ) {
 				return new MultiTableDeleteExecutor( walker );
 			}
 			else {
-				return new DeleteExecutor( walker, persister );
+				return new DeleteExecutor( walker );
 			}
 		}
 		else if ( walker.getStatementType() == HqlSqlTokenTypes.UPDATE ) {
 			final FromElement fromElement = walker.getFinalFromClause().getFromElement();
 			final Queryable persister = fromElement.getQueryable();
-			if ( persister.isMultiTable() ) {
-				// even here, if only properties mapped to the "base table" are referenced
-				// in the set and where clauses, this could be handled by the BasicDelegate.
-				// TODO : decide if it is better performance-wise to doAfterTransactionCompletion that check, or to simply use the MultiTableUpdateDelegate
+
+			if ( persister.isMultiTable() && affectsExtraTables( walker, persister ) ) {
 				return new MultiTableUpdateExecutor( walker );
 			}
+			else if ( persister.isMultiTable() && walker.getQuerySpaces().size() > 1 ) {
+				return new IdSubselectUpdateExecutor( walker );
+			}
 			else {
-				return new BasicExecutor( walker, persister );
+				return new SimpleUpdateExecutor( walker );
 			}
 		}
 		else if ( walker.getStatementType() == HqlSqlTokenTypes.INSERT ) {
-			return new BasicExecutor( walker, ( (InsertStatement) statement ).getIntoClause().getQueryable() );
+			return new InsertExecutor( walker );
 		}
 		else {
 			throw new QueryException( "Unexpected statement type" );
 		}
 	}
+
+	private static boolean affectsExtraTables(HqlSqlWalker walker, Queryable persister) {
+		String[] tableNames = persister.getConstraintOrderedTableNameClosure();
+		return IntStream.range( 0, tableNames.length ).filter(
```

Comment:
IMO it would be nicer to have `org.hibernate.hql.internal.ast.tree.AssignmentSpecification` expose the `tableNames` set instead and do the following:

```
Set<String> tables = new HashSet<>( 2 );
for ( AssignmentSpecification assign : walker.getAssignmentSpecifications() ) {
    tables.addAll( assign.getTableNames() );
    if ( tables.size() > 1 ) {
        return true;
    }
}
return false;
```

Reply:
I agree. I particularly dislike that `affectsTable()` takes a string.

But that's an orthogonal change not addressed in this issue.